### PR TITLE
bstone: init at 1.2.12

### LINKED ISSUE #232638: Package request: bstone
**Project description**

Unofficial source port for Blake Stone series (Aliens Of Gold and Planet Strike).

**Metadata**

* homepage URL:
* source URL: https://github.com/bibendovsky/bstone
* license: mit
* platforms: linux

### PR DESCRIPTION
**Pull request by keenanweaver**
## Description of changes

Adds bstone, closes https://github.com/NixOS/nixpkgs/issues/232638

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/bs/bstone/package.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  src = fetchFromGitHub {
+    owner = "bibendovsky";
+    repo = "bstone";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-wtW595cSoVTZaVykxOkJViNs3OmuIch9nA5s1SqwbJo=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    gcc12
```
Comment (msanft) [inline pkgs/by-name/bs/bstone/package.nix:23]: If this really needs GCC 12, maybe use `stdenv = gcc12Stdenv` instead

File: pkgs/by-name/bs/bstone/package.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    libGL
+    SDL2
+  ];
+
+  postInstall = ''
+    mkdir -p $out/bin
+    mv $out/bstone* $out/bin
+  '';
+
+  meta = with lib; {
```
Comment (msanft) [inline pkgs/by-name/bs/bstone/package.nix:36]:
```suggestion
  meta = {
```
nit: There's some [reasoning](https://github.com/NixOS/nixpkgs/pull/294383/files) for not going `with lib` on `meta`.

File: pkgs/by-name/bs/bstone/package.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    description = "Unofficial source port for the Blake Stone series";
+    homepage = "https://github.com/bibendovsky/bstone";
+    changelog = "https://github.com/bibendovsky/bstone/blob/${finalAttrs.src.rev}/CHANGELOG.md";
+    license = with licenses; [
```
Comment (msanft) [inline pkgs/by-name/bs/bstone/package.nix:40]:
```suggestion
    license = with lib.licenses; [
```

File: pkgs/by-name/bs/bstone/package.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      gpl2Plus # Original game source code
+      mit # BStone
+    ];
+    maintainers = with maintainers; [ keenanweaver ];
```
Comment (msanft) [inline pkgs/by-name/bs/bstone/package.nix:44]:
```suggestion
    maintainers = with lib.maintainers; [ keenanweaver ];
```

File: pkgs/by-name/bs/bstone/package.nix [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    mainProgram = "bstone";
+    platforms = platforms.linux; #TODO: macOS / Darwin support
```
Comment (msanft) [inline pkgs/by-name/bs/bstone/package.nix:46]:
```suggestion
    platforms = lib.platforms.linux; #TODO: macOS / Darwin support
```

File: pkgs/by-name/bs/bstone/package.nix (hunk 1 above)
Comment (keenanweaver) [reply pkgs/by-name/bs/bstone/package.nix:23]: Yeah, I did try this but for whatever reason it doesn't build unless gcc12 is in the build inputs. It's probably due to my ignorance, and I'm not sure how to fix it. This is the log using just `stdenv = gcc12Stdenv`

```
this derivation will be built:
  /nix/store/ih050wgid4qrva1gixfa6m71mahknm0d-bstone-1.2.12.drv
building '/nix/store/ih050wgid4qrva1gixfa6m71mahknm0d-bstone-1.2.12.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/apackvzn5f2q45zawgpfighn42zn7b8w-source
source root is source
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
fixing cmake files...
cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/share/doc/bstone_solution -DCMAKE_INSTALL_INFODIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/include -DCMAKE_INSTALL_SBINDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/ac1hb5dc2z4biwgy8mjrhlifffkkrvdq-gcc-wrapper-13.2.0/bin/strip -DCMAKE_RANLIB=/nix/store/ac1hb5dc2z4biwgy8mjrhlifffkkrvdq-gcc-wrapper-13.2.0/bin/ranlib -DCMAKE_AR=/nix/store/ac1hb5dc2z4biwgy8mjrhlifffkkrvdq-gcc-wrapper-13.2.0/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/a00cqsbns6q340xs0pfpf8sihddbkk0g-bstone-1.2.12
CMake Deprecation Warning at CMakeLists.txt:9 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at src/CMakeLists.txt:9 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The CXX compiler identification is GNU 13.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/ac1hb5dc2z4biwgy8mjrhlifffkkrvdq-gcc-wrapper-13.2.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [bstone] v1.2.12
CMake Deprecation Warning at cmake/FindSDL2W.cmake:41 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
Call Stack (most recent call first):
  src/CMakeLists.txt:46 (find_package)


-- [SDL2W] Version: 1.0.4
-- [SDL2W] No components were selected.
-- [SDL2W] Found CMake config.
-- [SDL2W] Libraries:
-- [SDL2W]     "SDL2::SDL2"
-- [SDL2W] Include directories:
-- [SDL2W]    "/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include"
-- [SDL2W]    "/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2"
-- [SDL2W]    "/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2"
-- [SDL2W]    "/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2"
-- [SDL2W] Found SDL version: 2.30.1
-- Found SDL2W: /nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include;/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2;/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2;/nix/store/3p1c76j1w90kwfb2wf8lzd54wzhdk5iw-SDL2-2.30.1-dev/include/SDL2 (found suitable version "2.30.1", minimum required is "2.0.4")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Configuring done (0.4s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_TESTING
    CMAKE_C_COMPILER
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_INSTALL_BINDIR
    CMAKE_INSTALL_DOCDIR
    CMAKE_INSTALL_INCLUDEDIR
    CMAKE_INSTALL_INFODIR
    CMAKE_INSTALL_LIBDIR
    CMAKE_INSTALL_LIBEXECDIR
    CMAKE_INSTALL_LOCALEDIR
    CMAKE_INSTALL_MANDIR
    CMAKE_INSTALL_OLDINCLUDEDIR
    CMAKE_INSTALL_SBINDIR
    CMAKE_POLICY_DEFAULT_CMP0025


-- Build files have been written to: /build/source/build
cmake: enabled parallel building
cmake: enabled parallel installing
Running phase: buildPhase
build flags: -j16 SHELL=/nix/store/a1s263pmsci9zykm5xcdf7x9rv26w6d5-bash-5.2p26/bin/bash
[  1%] Building CXX object src/CMakeFiles/bstone.dir/dosbox/dbopl.cpp.o
[  1%] Building CXX object src/CMakeFiles/bstone.dir/lib/xbrz/xbrz.cpp.o
[  1%] Building CXX object src/CMakeFiles/bstone.dir/bstone_offset_2d.cpp.o
[  2%] Building CXX object src/CMakeFiles/bstone.dir/bstone_extent_2d.cpp.o
[  3%] Building CXX object src/CMakeFiles/bstone.dir/bstone_hw_shader_registry.cpp.o
[  3%] Building CXX object src/CMakeFiles/bstone.dir/bstone_hw_texture_mgr.cpp.o
[  4%] Building CXX object src/CMakeFiles/bstone.dir/bstone_rect_2d.cpp.o
[  5%] Building CXX object src/CMakeFiles/bstone.dir/bstone_missing_sprite_64x64_image.cpp.o
[  7%] Building CXX object src/CMakeFiles/bstone.dir/bstone_ren_3d_mgr.cpp.o
[  7%] Building CXX object src/CMakeFiles/bstone.dir/bstone_missing_wall_64x64_image.cpp.o
[  7%] Building CXX object src/CMakeFiles/bstone.dir/bstone_sdl_exception.cpp.o
[  7%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_api.cpp.o
[  8%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_context.cpp.o
[  9%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_utils.cpp.o
[ 10%] Building CXX object src/CMakeFiles/bstone.dir/bstone_ren_3d_cmd_buffer.cpp.o
[ 11%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_buffer.cpp.o
[ 12%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_extension_mgr.cpp.o
[ 12%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_error.cpp.o
[ 13%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl.cpp.o
[ 14%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_utils.cpp.o
[ 14%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_sampler.cpp.o
[ 15%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_sampler_mgr.cpp.o
[ 16%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader.cpp.o
[ 17%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage.cpp.o
[ 17%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage_mgr.cpp.o
[ 18%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_var.cpp.o
[ 19%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_texture_2d.cpp.o
[ 19%] Building CXX object src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_texture_mgr.cpp.o
In file included from /build/source/src/bstone_ren_3d_shader_stage.h:24,
                 from /build/source/src/bstone_detail_ren_3d_gl_shader_stage.h:20,
                 from /build/source/src/bstone_detail_ren_3d_gl_shader_stage.cpp:16:
/build/source/src/bstone_ren_3d_shader_var.h:91:28: error: 'int32_t' in namespace 'std' does not name a type
   91 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
/build/source/src/bstone_ren_3d_shader_var.h:211:28: error: 'int32_t' in namespace 'std' does not name a type
  211 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
In file included from /build/source/src/bstone_ren_3d_shader_stage.h:24,
                 from /build/source/src/bstone_detail_ren_3d_gl_shader_stage_mgr.h:20,
                 from /build/source/src/bstone_detail_ren_3d_gl_shader_stage_mgr.cpp:14:
/build/source/src/bstone_ren_3d_shader_var.h:91:28: error: 'int32_t' in namespace 'std' does not name a type
   91 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
/build/source/src/bstone_ren_3d_shader_var.h:211:28: error: 'int32_t' in namespace 'std' does not name a type
  211 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
In file included from /build/source/src/bstone_detail_ren_3d_gl_shader_var.h:20,
                 from /build/source/src/bstone_detail_ren_3d_gl_shader_var.cpp:16:
/build/source/src/bstone_ren_3d_shader_var.h:91:28: error: 'int32_t' in namespace 'std' does not name a type
   91 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
/build/source/src/bstone_ren_3d_shader_var.h:211:28: error: 'int32_t' in namespace 'std' does not name a type
  211 |                 const std::int32_t value) = 0;
      |                            ^~~~~~~
make[2]: *** [src/CMakeFiles/bstone.dir/build.make:412: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage_mgr.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src/CMakeFiles/bstone.dir/build.make:398: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage.cpp.o] Error 1
make[2]: *** [src/CMakeFiles/bstone.dir/build.make:426: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_var.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:98: src/CMakeFiles/bstone.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
error: builder for '/nix/store/ih050wgid4qrva1gixfa6m71mahknm0d-bstone-1.2.12.drv' failed with exit code 2;
       last 25 log lines:
       >   211 |                 const std::int32_t value) = 0;
       >       |                            ^~~~~~~
       > In file included from /build/source/src/bstone_ren_3d_shader_stage.h:24,
       >                  from /build/source/src/bstone_detail_ren_3d_gl_shader_stage_mgr.h:20,
       >                  from /build/source/src/bstone_detail_ren_3d_gl_shader_stage_mgr.cpp:14:
       > /build/source/src/bstone_ren_3d_shader_var.h:91:28: error: 'int32_t' in namespace 'std' does not name a type
       >    91 |                 const std::int32_t value) = 0;
       >       |                            ^~~~~~~
       > /build/source/src/bstone_ren_3d_shader_var.h:211:28: error: 'int32_t' in namespace 'std' does not name a type
       >   211 |                 const std::int32_t value) = 0;
       >       |                            ^~~~~~~
       > In file included from /build/source/src/bstone_detail_ren_3d_gl_shader_var.h:20,
       >                  from /build/source/src/bstone_detail_ren_3d_gl_shader_var.cpp:16:
       > /build/source/src/bstone_ren_3d_shader_var.h:91:28: error: 'int32_t' in namespace 'std' does not name a type
       >    91 |                 const std::int32_t value) = 0;
       >       |                            ^~~~~~~
       > /build/source/src/bstone_ren_3d_shader_var.h:211:28: error: 'int32_t' in namespace 'std' does not name a type
       >   211 |                 const std::int32_t value) = 0;
       >       |                            ^~~~~~~
       > make[2]: *** [src/CMakeFiles/bstone.dir/build.make:412: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage_mgr.cpp.o] Error 1
       > make[2]: *** Waiting for unfinished jobs....
       > make[2]: *** [src/CMakeFiles/bstone.dir/build.make:398: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_stage.cpp.o] Error 1
       > make[2]: *** [src/CMakeFiles/bstone.dir/build.make:426: src/CMakeFiles/bstone.dir/bstone_detail_ren_3d_gl_shader_var.cpp.o] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:98: src/CMakeFiles/bstone.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/ih050wgid4qrva1gixfa6m71mahknm0d-bstone-1.2.12.drv'.
```

Comment (msanft) [reply pkgs/by-name/bs/bstone/package.nix:23]: Hmm, it built just fine for me. Note that you don't need to set `stdenv = gcc12Stdenv` in the attribute set of `stdenv.mkDerivation`, but in a let-block before that, like so:

```nix
let
  stdenv = gcc12Stdenv;
in 
stdenv.mkDerivation
``

Alternatively, you could also entirely substitute `stdenv` with `gcc12Stdenv`, which would even be preferable imo. Sorry for the confusion due to my first suggestion.

Comment (keenanweaver) [reply pkgs/by-name/bs/bstone/package.nix]: That was it. You are correct! I've made the changes now.

> Alternatively, you could also entirely substitute `stdenv` with `gcc12Stdenv`

I'm not a stickler for style too much; do you know which would be preferred?

**Review by msanft — APPROVED** [review]
Thank you!

**Comment by keenanweaver** [comment]
> Thank you!

Thank you for your time and wisdom!

**Pull request merged**